### DISCUSSION TITLE
Parsing Korg KARMA Sys-Ex

### DISCUSSION THREAD

**Discussion by thethirdapple** [RomanKubiak/ctrlr#547] [Q&A] [answered]
Greetings,

Starting a new thread regarding the parsing of Korg Karma Sysex strings when switching programs / banks... 

**NOT URGENT** but for future reference and eventual panel functionality.

So while it is possible to switch banks and program presets with a rather simple a straightforward external sysex from a panel such as :
B0 20 xx [bank]
C0 xx [program]

I've taken note that the on board controls returns a super long sysex string within a single F0 ... F7 message. Which seems to contain a bunch of internal settings which pertain to the preset. This is returned when using the on board bank buttons but not when switching bank / programs with the Ctrlr panel???

So I am wondering if it is possible to parse the single string into it components so as to "set" sliders within the Ctrlr panel?

**Side note**, if interested: I have been cataloging, into a spreadsheet,  ALL the sysex messages to control the synth and the KARMA functions and happy to post here or goto , [KARMA-Lab Forums](https://forums.karma-lab.com/forum/showthread.php?t=22078), where I've been posting updates as the spreadsheet progresses and  hopefully Stephen will call out any errors I'm making... so far so good.

Below is an example of the returned sysex from the keyboard when program / banks are switched using the on board controls

Peace

> F0 42 30 5D 40 00 00 44 61 72 6B 20 52 26 00 42 20 42 61 73 73 20 28 20 20 24 44 0C 43 04 45 47 14 5E 03 00 64 55 00 32 5A 1A 00 42 40 00 00 00 40 00 7F 7F 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 7F 00 7F 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 7F 7F 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 7F 7F 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 10 00 7F 7F 22 64 01 14 20 00 27 05 14 64 49 00 14 14 00 1B 27 28 31 40 00 00 00 02 5A 00 14 00 00 08 00 00 79 04 13 31 00 00 00 00 00 35 40 00 00 00 7F 7F 00 7F 13 02 00 06 03 08 18 00 00 00 10 00 00 00 62 43 00 5E 00 00 17 01 43 01 53 00 14 41 01 15 00 62 01 25 00 00 06 00 7C 00 3B 56 00 32 40 00 32 00 7F 3B 00 61 32 40 7F 7F 7F 00 00 00 00 41 7F 00 00 00 00 00 41 7F 00 00 00 00 00 41 7F 00 00 00 00 41 20 7F 00 00 00 00 7F 00 00 00 00 00 00 00 00 00 02 00 7F 00 00 00 00 00 10 00 00 00 00 7F 00 00 00 00 00 00 00 00 00 00 01 7F 00 00 00 00 00 00 08 00 00 00 7F 00 00 00 40 00 00 00 00 00 00 7F 00 00 00 00 00 00 00 00 04 00 00 7F 00 00 00 00 20 00 00 00 00 00 7F 00 40 00 00 00 00 00 30 7F 3F 7F 7F 7F 7F 7F 7F 2B 7F 7F 7F 7F 7F 7F 7F 7F 7E 2E 7F 7F 7F 7F 7F 7F 7D 7F 29 7F 7F 7F 7F 7F 03 7F 7F 4E 4A 46 4C 00 00 00 00 0F 10 10 00 34 04 00 00 30 7F 00 00 00 40 00 00 30 31 00 00 2F 45 7F 00 40 00 00 00 7F 00 00 00 06 01 00 00 32 00 02 00 00 18 03 00 00 00 40 0D 00 00 2A 04 01 00 00 00 00 00 00 00 10 00 00 00 01 11 00 00 01 00 0E 00 00 07 07 00 00 22 40 7F 00 00 00 7F 00 44 00 00 7F 00 00 00 7F 08 00 00 00 7F 00 00 00 54 47 2D 7F 00 7F 00 7F 02 00 00 00 00 00 00 00 00 2D 01 08 00 00 00 01 02 00 10 11 12 04 4C 40 08 00 58 42 4D 26 00 0F 40 00 0E 00 00 00 00 4A 0C 00 7F 00 49 00 7F 00 00 01 01 7F 00 46 00 00 00 00 00 0E 00 13 00 00 00 46 00 00 00 00 0E 00 04 13 00 7F 00 00 00 14 00 5A 0A 00 26 60 00 00 08 02 05 02 74 00 00 34 00 00 11 20 11 00 00 63 00 12 0F 00 18 11 11 12 00 19 18 1C 0F 25 11 00 00 00 00 00 00 00 00 00 00 00 00 12 00 11 00 00 00 00 00 00 00 00 00 00 20 00 0C 0F 01 29 7F 47 00 01 1E 0D 14 00 05 06 40 0A 0F 00 0F 13 3E 68 00 40 24 7F 23 11 00 00 00 00 14 00 14 00 00 00 00 63 44 60 5A 4F 05 0A 00 50 0F 00 0F 00 14 00 20 00 00 14 00 20 71 00 20 40 0E 00 7F 7F 01 16 0C 00 7F 00 22 00 7F 00 00 01 01 7F 00 46 00 00 00 00 00 0E 00 13 00 00 00 46 00 00 00 00 0E 00 04 13 00 7F 00 00 00 14 00 00 0A 00 26 5F 00 00 08 02 12 02 74 00 00 34 00 00 11 14 11 00 00 63 00 00 0F 00 18 11 11 1C 04 00 18 3C 0F 1B 05 11 00 00 00 00 00 00 00 00 00 00 00 12 00 11 00 00 00 00 00 00 00 00 00 00 00 00 00 00 63 3C 44 63 00 23 3B 38 14 14 05 06 05 76 0F 7F 0F 31 54 40 00 55 18 7F 0F 11 00 00 00 00 14 00 14 00 00 00 00 63 44 60 63 44 05 0A 00 00 50 00 0F 00 14 00 20 00 00 11 00 17 71 00 00 40 0E 00 7F 7F 00 00 F7

**Comment by dnaldoog** [comment, marked as answer]
Here is what I think is probably correct parsing of the data. It is adapted from my M3R panel &#8212; so some small details may be incorrect.

[parse Korg Karma sysex_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/11057873/parse.Korg.Karma.sysex_1_0.zip)

The data `sysexString` is the string above in your post.

<pre>
parseKorgData = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local offset = 6 -- ignore KORG Header
    local nameSize = 16

    local m = MemoryBlock()
    m:loadFromHexString(sysexString)
    console(string.format("Initial Dump Size=%d bytes", m:getSize()))
    local numberofpatchesindump = m:getSize() -- this will be something different (like 128 etc)

    local sizeOfLoop = m:getSize() - 1 - 4 - 7 --[[ always getSize()-1,-4 extra nul bytes,-7 header+eox--]]
    local accrue = MemoryBlock()
    for i = offset, sizeOfLoop, 8 do
        local set = MemoryBlock(m:getRange(i, 8))
        local res = convert827(set)
        accrue:append(res)
    end

    for i = 0, accrue:getSize() - 1, numberofpatchesindump do
        console(accrue:getRange(i, nameSize):toString()) -- extract all names
    end
    console(string.format("Actual Korg Data: %s\n%d bytes", accrue:toHexString(1), accrue:getSize()))
end


function convert827(m) -- take a set of 8 bytes (first byte contains) msbit data of each subsequent byte
    local t = {}
    assert(m:getSize() == 8, "size=" .. m:getSize())
    local b = {1, 2, 4, 8, 16, 32, 64}
    local msb = m:getByte(0) --
    for i = 1, m:getSize() - 1 do -- start at the second byte
        local mask = b[i]
        local band = bit.rshift(bit.band(msb, mask), i - 1)
        local addmsb = m:getByte(i) + (band * 128)
        table.insert(t, addmsb)
    end
    local result = MemoryBlock()
    result:createFromTable(t)
    assert(result:getSize() == 7)
    return result -- MEMORYBLOCK
end
</pre>

**Reply by thethirdapple** [reply]
Thank you, I will try this out in the morning… see if I am able to make sense of it.
Truly awesome!!!

**Reply by dnaldoog** [reply]
I'm not sure about this line:
`local sizeOfLoop = m:getSize() - 1 - 4 - 7 --[[ always getSize()-1,-4 extra nul bytes,-7 header+eox--]]`
It could be
`local sizeOfLoop = m:getSize() - 1 - 2 - 7 --[[ always getSize()-1,-2 extra nul bytes,-7 header+eox--]]`
or even
`local sizeOfLoop = m:getSize() - 1 - 0 - 7 --[[ always getSize()-1,-0 extra nul bytes,-7 header+eox--]]`

**Reply by thethirdapple** [reply]
Looks to be functioning properly as written in the zipped panel... Really excited to go down this path and learn way more than anticipated!!!
![Screen Shot 2023-03-24 at 9 15 01 AM](https://user-images.githubusercontent.com/125695472/227531252-7cf4458a-ea34-47f2-8aad-a5aae06d08a4.png)

**Reply by dnaldoog** [reply]
The next step would be to map each parameter in the manual to each byte in the message and update the panel. 

The first 16 bytes are the name in ASCII `44 61 72 6b 20 52 26 42 20 42 61 73 73 20 20 20` _Dark R&B Bass_

Sometimes there will be bitpacking of several values - say 8 modulators have on/off stored in a single byte, so it will be a challenge mapping everything out for 800+ values! 

I always build a lookup table that tells me how to break down the data. For example in the Korg M3R I have a lookup table (excerpt below):

The `key`/`hash` to each  table record is the modulator name in the panel, `byte` is which byte in `accrue` MemoryBlock() and `mask` is used to extract bitfields (0 = use the whole byte). From Memory some parameters on the M3R needed to be reduced/incremented by 1, hence the `add` which is usually 0 as you can see.
<pre>
    t={
	...
    ["CONTROL CHANGE1"] = {byte = 45, mask = 4, rs = 2, label = "Ctrl. Chge", max = 1, min=0},
    ["CONTROL CHANGE2"] = {byte = 56, mask = 4, rs = 2, label = "Ctrl. Chge", max = 1, min=0},
    ["CONTROL CHANGE3"] = {byte = 67, mask = 4, rs = 2, label = "Ctrl. Chge", max = 1, min=0},
    ["CONTROL CHANGE4"] = {byte = 78, mask = 4, rs = 2, label = "Ctrl. Chge", max = 1, min=0},
    ["CONTROL CHANGE5"] = {byte = 89, mask = 4, rs = 2, label = "Ctrl. Chge", max = 1, min=0},
    ["CONTROL CHANGE6"] = {byte = 100, mask = 4, rs = 2, label = "Ctrl. Chge", max = 1, min=0},
    ["CONTROL CHANGE7"] = {byte = 111, mask = 4, rs = 2, label = "Ctrl. Chge", max = 1, min=0},
    ["CONTROL CHANGE8"] = {byte = 122, mask = 4, rs = 2, label = "Ctrl. Chge", max = 1, min=0},
    ["DAMPER1"] = {byte = 45, mask = 8, rs = 3, label = "Damper", max = 1, min=0},
    ["DAMPER2"] = {byte = 56, mask = 8, rs = 3, label = "Damper", max = 1, min=0},
    ["DAMPER3"] = {byte = 67, mask = 8, rs = 3, label = "Damper", max = 1, min=0},
    ["DAMPER4"] = {byte = 78, mask = 8, rs = 3, label = "Damper", max = 1, min=0},
    ["DAMPER5"] = {byte = 89, mask = 8, rs = 3, label = "Damper", max = 1, min=0},
    ["DAMPER6"] = {byte = 100, mask = 8, rs = 3, label = "Damper", max = 1, min=0},
    ["DAMPER7"] = {byte = 111, mask = 8, rs = 3, label = "Damper", max = 1, min=0},
    ["DAMPER8"] = {byte = 122, mask = 8, rs = 3, label = "Damper", max = 1, min=0},
    ["DETUNE1"] = {byte = 39, mask = 255, rs = 0, label = "Detune", max = 50, min = -50,add=0}, ...
	}
	-- Once you have your data from a dump, loop through the table and assign values

	for k, a in pairs(t) do
        local value = bit.rshift(bit.band(accrue:getByte(a.byte), a.mask), a.rs) + a.add
        panel:getcomponent(k):setValue(signed8bitChar(value), true) -- set the modulator value to the byte extracted from the dump
	 end
		
		--The signed8bitChar function converts bytes to negatives**
		
function signed8bitChar(val)
val=val < 128 and val or (val -256 )
return val
end
</pre>
If you see a FF in accrue MemoryBlock data, I am _guessing_ it means assign -1 to your modulator (at least, from memory this is the case on the M3R)
eg `0xff = -1,0x80=-128, 0x7f=127`**

**Reply by thethirdapple** [reply]
Sweetness !!! As always the pre-emptive answers to questions not yet posed are appreciated. I am not quite ready to implement this but having it out there (here) reminds me of all the interrelated aspects I hope take into account in an eventual build I can share... Currently in the depth of creating a complete SysEx message spreadsheet, to facilitate any future look up tables and dynamic assignments of sliders.
